storage/Devices/PartitionImpl.cc

```diff
@@ -499,16 +580,71 @@ namespace storage
 
 	// TODO
 
-	if (get_id() > 255 || !is_msdos(partition_table))
-	    return;
+	if (is_msdos(partition_table))
+	{
+	    // Note: The type option is not available in upstream parted.
 
-	string cmd_line = PARTEDBIN " --script " + quote(partitionable->get_name()) + " set " +
-	    to_string(get_number()) + " type " + to_string(get_id());
-	cout << cmd_line << endl;
+	    string cmd_line = PARTEDBIN " --script " + quote(partitionable->get_name()) + " set " +
+		to_string(get_number()) + " type " + to_string(get_id());
+	    cout << cmd_line << endl;
 
-	SystemCmd cmd(cmd_line);
-	if (cmd.retcode() != 0)
-	    ST_THROW(Exception("set partition id failed"));
+	    SystemCmd cmd(cmd_line);
+	    if (cmd.retcode() != 0)
+		ST_THROW(Exception("set partition id failed"));
+	}
+	else
+	{
+	    string cmd_line = PARTEDBIN " --script " + quote(partitionable->get_name()) + " set " +
+		to_string(get_number()) + " ";
+
+	    switch (get_id())
+	    {
+		case ID_LINUX:
+		    // this is tricky but parted has no clearer way
+		    cmd_line += "lvm off";
+		    break;
+
+		case ID_SWAP:
+		    cmd_line += "swap on";
+		    // TODO improve parted
+		    ST_THROW(Exception("set swap not supported by parted"));
+		    break;
+
+		case ID_LVM:
+		    cmd_line += "lvm on";
+		    break;
+
+		case ID_RAID:
+		    cmd_line += "raid on";
+		    break;
+
+		case ID_ESP:
+		    cmd_line += "esp on";
+		    break;
+
+		case ID_BIOS_BOOT:
+		    cmd_line += "bios_grub on";
+		    break;
+
+		case ID_PREP:
+		    cmd_line += "prep on";
+		    break;
+
+		case ID_WINDOWS_BASIC_DATA:
+		    cmd_line += "msftdata";
+		    break;
+
+		case ID_MICROSOFT_RESERVED:
+		    cmd_line += "msftres";
+		    break;
+	    }
+
+	    cout << cmd_line << endl;
+
+	    SystemCmd cmd(cmd_line);
+	    if (cmd.retcode() != 0)
+		ST_THROW(Exception("set partition id failed"));
```

Comment:
NP: isn't it redundant to start and finish both branches of the script with basically the same lines of code (preparing cmd_line and executing it respectively)?